docs/decisions/0024-planner-telemetry-enhancement.md

```diff
@@ -0,0 +1,366 @@
+
+1. SK is already configured to propagate token usage information from connectors via `ModelResult`. We simply need to extend the list of items that get propagated, such as model information.
+2. Currently, SK does not have a method for passing function information down to the connector level. Although we considered using [baggage](https://opentelemetry.io/docs/concepts/signals/baggage/#:~:text=In%20OpenTelemetry%2C%20Baggage%20is%20contextual%20information%20that%E2%80%99s%20passed,available%20to%20any%20span%20created%20within%20that%20trace.) as a means of propagating information downward, experts from the OpenTelemetry team advised against this approach due to security concerns.
+
+With the bottom-up approach, we need to add model information to `IResultBase`:
+
+```csharp
+/// <summary>
+/// Interface for model results
+/// </summary>
+public interface IResultBase
+{
+  /// <summary>
+  /// Model name or Id
+  /// </summary>
+  string ModelId { get; }
+
+  /// <summary>
+  /// Gets the model result data.
+  /// </summary>
+  ModelResult ModelResult { get; }
+}
+```
+
+We also need to propagate the model Id up via KernelResult:
+
+```csharp
+// In AIFunctionResultExtensions.cs
+...
+/// <summary>
+/// Function model id key for <see cref="ModelResult"/> records.
+/// </summary>
+public const string ModelIdKey = "ModelId";
+...
+/// <summary>
+/// Returns the model Id from <see cref="FunctionResult"/> metadata.
+/// </summary>
+/// <param name="result">Instance of <see cref="FunctionResult"/> class.</param>
+public static string? GetModelId(this FunctionResult result)
+{
+  if (result.TryGetMetadataValue(ModelIdKey, out string? modelId))
+  {
+    return modelId;
+  }
+
+  return null;
+}
+
+
+// In KernelFunctionFromPrompt.cs
+...
+result.Metadata.Add(AIFunctionResultExtensions.ModelIdKey, completionResults[0].ModelId);
+...
+return result;
+```
+
+A kernel function will then retrieve the information by doing the following:
+
+```csharp
+// Note that not all services support usage details.
+if (result.GetModelId() is null)
+{
+  logger.LogInformation("Model id not found in function result.");
+  return;
+}
+var modelId = result.GetModelId();
+
+if (result.GetModelResults() is null)
+{
+  logger.LogInformation("Model results not found in function result.");
+  return;
+}
+var modelResult = result.GetModelResults().FirstOrDefault();
+var modelResultJson = modelResult?.GetJsonResult();
+
+var promptTokens = 0;
+var completionTokens = 0;
+try
+{
+  var tokenUsage = modelResultJson.GetValueOrDefault().GetProperty("Usage");
+  promptTokens = tokenUsage.GetProperty("PromptTokens").GetInt32();
+  completionTokens = tokenUsage.GetProperty("CompletionTokens").GetInt32();
+}
+catch (Exception ex) when (ex is KeyNotFoundException)
+{
+  logger.LogInformation("Usage details not found in model result.");
+  return;
+}
+catch (Exception ex)
+{
+  logger.LogError(ex, "Error while parsing usage details from model result.");
+  throw;
+}
```

Comment:
How are the values being used? This snippet looks incomplete.

Reply:
More implementation details added.